ophyd/_dispatch.py

```diff
@@ -0,0 +1,134 @@
+import time
+import functools
+import queue
+import threading
+
+
+class _CallbackThread(threading.Thread):
+    'A queue-based callback dispatcher thread'
+
+    def __init__(self, name, *, dispatcher, **kwargs):
+        super().__init__(name=name, **kwargs)
+        self.daemon = True
+        self.stop_event = dispatcher._stop_event
+        self.context = dispatcher.context
+        self._timeout = dispatcher._timeout
+        self.logger = dispatcher.logger
+        self.queue = queue.Queue()
+        self.current_callback = None
+
+    def __repr__(self):
+        return '<{} qsize={}>'.format(self.__class__.__name__,
+                                      self.queue.qsize())
+
+    def run(self):
+        '''The dispatcher itself'''
+        self.attach_context(self.context)
+        self.logger.debug('Callback thread %s started', self.name)
+        while not self.stop_event.is_set():
+            try:
+                callback, args, kwargs = self.queue.get(True, self._timeout)
+            except queue.Empty:
+                ...
+            else:
+                try:
+                    self.current_callback = (callback.__name__, kwargs.get('pvname'))
+                    callback(*args, **kwargs)
+                except Exception as ex:
+                    self.logger.exception(
+                        'Exception occurred during callback %r', callback
+                    )
+
+        self.logger.debug('Callback thread %s exiting', self.name)
+        self.detach_context()
+
+    def attach_context(self, context):
+        self.logger.debug('Callback thread %s attaching to context %s',
+                          self.name, self.context)
```

Comment:
Is this missing

```py
self.context = context
```

Reply:
It surprised me that `__init__` takes a `dispatcher` and picks up a lot of its attributes, but `attach_context` takes a `context` directly but hangs onto the old dispatcher's other attributes.

Reply:
I could make up excuses here, but it was just sloppy. All things picked up from `dispatcher` previously are now kwargs.